src/dynamic_boundary_conditions/sea_level_rise_data.py

```diff
@@ -3,106 +3,95 @@
-
-    sea_level_rise_projections_nz.rename(
-        columns={'siteId': 'Site_ID', 'year': 'Projected_Year', 'p17': 'SLR_p17_confidence_interval_meters',
-                 'p50': 'SLR_p50_confidence_interval_meters', 'p83': 'SLR_p83_confidence_interval_meters',
-                 'lon': 'Longitude', 'lat': 'Latitude', 'measurementName': 'Scenario_(VLM_with_confidence)', },
-        inplace=True)
-
-    return sea_level_rise_projections_nz
-
-
-def unique_locations(sea_level_rise_projections_nz: gpd.geodataframe.GeoDataFrame):
-    """Finds the unique lat lon coordinates (site locations) and appends them to list and returns that list"""
-    locations_lat_lon = []
-    # Acquiring unique pairs
-    unique_location = sea_level_rise_projections_nz.geometry.unique()
-    # Adding pairs to list of dictionarys to map to later on
-    print('Finding unique locations...')
-    for i in unique_location:
-        locations_value_pair = {'lat': i.y, 'lon': i.x}
-        locations_lat_lon.append(locations_value_pair)
-    return locations_lat_lon
+            try:
+                print('LOADED CSV file by pyarrow: ', filename)
+                #  pyarrow is faster, need Pandas 1.4, released in January 2022
+                df = pd.read_csv(file_path + filename, engine='pyarrow', dtype={'siteId': int})
+            except:
+                print('LOADED CSV file: ', filename)
+                df = pd.read_csv(file_path + filename, dtype={'siteId': int})
+            df["Region"] = filename[24:-4]
+            df_list.append(df)
+    return pd.concat(df_list, ignore_index=True, axis=0)
+
+
+def get_nearest_coordinate(df: pd.DataFrame, target: tuple) -> tuple:
+    """ find the nearest coordinate to target position in the input dataframe."""
+    # get unique coordinate
+    df_site = df.copy()[['lat', 'lon']].drop_duplicates(subset=['lat', 'lon'])
+    # protect input boundary for haversine function
+    assert -90 <= target[0] <= 90, "Latitude is out of range [-90, 90]"
+    assert -180 <= target[0] <= 180, "Longitude is out of range [-180, 180]"
+    # calculate distance from target coordinate
+    df_site['distance'] = df_site.apply(lambda x: (haversine(target, (x['lat'], x['lon']))), axis=1)
+    # sort by distance
+    df_site = df_site.sort_values(by=['distance'], ascending=True).reset_index(drop=True)
+    print("Nearest coordinates: ({}, {})".format(df_site.iloc[0]['lat'], df_site.iloc[0]['lon']))
+    return df_site.iloc[0]['lat'], df_site.iloc[0]['lon']
+
+
+def gen_slr_data(df: pd.DataFrame, target: tuple, crs: str = "epsg:2193") -> gpd.GeoDataFrame:
+    """ generate geopandas dataframe that the coordinate is equal to target position."""
+    # filter the coordinates based on target coordinate.
+    df = df.loc[(df['lat'] == target[0]) & (df['lon'] == target[1])]
+    # convert to geopandas dataframe from pandas dataframe
+    gdf = gpd.GeoDataFrame(df, crs=crs, geometry=gpd.points_from_xy(df.lon, df.lat))
```

Comment:
I like the work around for using `pyarrow` instead of `geopandas`, and creating a `DataFrame` that can be converted to a `GeoDataFrame`. I think it would be nice if we could have used `GeoDataFrame` but the perfromance benefits are so huge using `pyarrow` that we can't ignore it.

Reply:
I think the efficiency of the geopandas dataframe is slightly lower than the vanilla pandas dataframe. The cost is acceptable in general. But the cost of converting from dataframe to geodataframe is pretty high, which is one of the significant time consumption in the original `sea_level_rise_data.py` script. So if the data to process are geodataframe compatible, it is OK to use geodataframe to load and process data. But if the data are not for the geodataframe, we should be aware of the high cost of converting the dataframe and only convert the dataframe when we have to.